Build with sandbox sources

rehno-lindeque opened the pull request:
Build the platform from sources using the more robust strategy I mentioned in https://github.com/elm-lang/elm-compiler/issues/995.

(Side note)
--------------
The only trouble I seem to be having is that I don't have `elm-make` in my `$PATH` by the time that I need to build `elm-reactor`:

```
Build log ( /home/rehno/projects/development/elm-platform/installers/Elm-Platform/0.15.1/.cabal-sandbox/logs/elm-reactor-0.3.2.log ):
Configuring elm-reactor-0.3.2...
Building elm-reactor-0.3.2...
Preprocessing executable 'elm-reactor' for elm-reactor-0.3.2...
Linking dist/dist-sandbox-f78d21b8/build/elm-reactor/elm-reactor ...
Custom build step: creating and collecting all static resources
setup: elm-make: does not exist
cabal: Error: some packages failed to install:
elm-reactor-0.3.2 failed during the building phase. The exception was:
ExitFailure 1
```

I'm not sure if this is actually a problem, could you try this install on your local machine @evancz ? I also created a patch to make sure that `elm-reactor` is built last if necessary: https://github.com/rehno-lindeque/elm-platform/compare/build-with-sandboxsources...build-elm-reactor

I was hoping that I could just temporarily change my `PATH` by running the command as
```bash
 PATH=./Elm-Platform/0.15.1/bin:$PATH runhaskell BuildFromSource.hs 0.15.1
```
but it seems as though `Setup.hs` must start on a fresh environment because that didn't do the trick for me.

Anyway, I'm about 95% sure this PR should do the trick (perhaps with https://github.com/rehno-lindeque/elm-platform/compare/build-with-sandboxsources...build-elm-reactor added on).

rehno-lindeque commented:
Oops, just noticed that I added the `--quiet` flag on a different git command than I was intending.

rehno-lindeque commented:
Ok, that's it - please let me know if it looks ok.

Apanatshka commented:
LGTM :+1: 
I noticed only a few days ago that doing something like this made builds easier. I did it like the following, but I didn't run into the elm-reactor post-build dependency issue because didn't try to compiler elm-reactor. If it wasn't for that, you could do it like I did and not have to fiddle with source-directories. 

```
cabal --ghc-options=-XFlexibleContexts\
      --sandbox-config-file=cabal.sandbox.config\
      install\
        elm-compiler/elm-compiler.cabal\
        elm-make/elm-make.cabal\
        elm-package/elm-package.cabal
```

rehno-lindeque commented:
Oh interesting! I didn't know you could just list cabal files that way. I'll see if I have time to update this PR tonight, that should make for a pleasingly tiny amount of code.

rehno-lindeque commented:
Sorry, I don't think I'm going to get to this soon, I have to prioritize a bit. However, the current PR is almost certainly an improvement over the previous state of this file (since the current strategy doesn't build at the moment I believe). 

I believe it is ok to merge as is.

evancz commented:
What do you think of the ideas listed in #60 about using `cabal freeze` everywhere once we have discovered a set of packages that works?

evancz commented:
@jvoigtlaender, is your comment https://github.com/elm-lang/elm-platform/issues/41#issuecomment-132195920 suggesting that this will close #41? If so, I will work on getting this plus `cabal freeze` in place for folks who are building from source.

jvoigtlaender commented:
I cannot say for certain whether https://github.com/elm-lang/elm-platform/issues/41#issuecomment-132195920 will close that issue. On my machine, I succeeded installing all the Elm tools using the fixed version numbers recorded in https://www.stackage.org/lts-3.1/cabal.config (by dropping that file into the project directories).

Fixing version numbers is generally the way to go, I think. Whether you do that by relying on the https://www.stackage.org/lts/cabal.config files or by trying to work out compatible versions yourself on your own machine and then do a `cabal freeze` is a question that I think will mainly affect your own comfort, not that of others ultimately installing from the repositories.

evancz commented:
I messed around with the add-source thing, but it seemed to mean that build times are always maximally bad. Like, if I just want to run `cabal build` on something, I need to install everything it depends on every time.

I don't think it's worth it to introduce such large delays. I am working on using a makefile to document all the tricks I do in my workflow and make it easier for others to do.

evancz commented:
I am going with the `cabal.config` approach where we just lock down all the dependencies for each release. I made a trial of this approach in https://github.com/elm-lang/elm-platform/tree/master/versions/0.15.1 and I think it'll do better than this PR in avoiding dependency stuff.

It's not yet clear to me if it will build with GHC 7.10 but I think I am okay saying to folks "You must use GHC 7.8" as long as we know that is true.

evancz commented:
Thanks everyone for working on getting this more reliable! We'll see if the `cabal.config` thing works out :)

rehno-lindeque commented:
No problem! I should probably mention @Apanatshka's technique would work without the cabal build issue. It doesn't matter though, anything that solves it for people just starting out with the source will be perfect (`cabal.config` ought to do the job).

jvoigtlaender commented:
Actually, I suspect the workflow from the build script (without `cabal add-source`) is not going to work *for people developing*. For people who only want to install Elm once from source, it should work, but for people trying to set up an environment for developing the tools, it probably won't (as has been observed in other issues).

Here is why:
* Say I set this up once. This will, among other things, build some version of `elm-compiler` and some version of `elm-package`. Say `elm-compiler-0.15.1` and `elm-package-0.5.1`. The latter needs the former, so when it is built, it is *linked against the `elm-compiler-0.15.1` package built and installed into the sandbox previously`.
* Say I now, as a developer, change something in the `elm-compiler` code. I rebuild the package after the change. That installs a new version (but with same version number 0.15.1) of `elm-compiler` into the sandbox. This *breaks* the `elm-package` built, because that one was linked against a now not anymore existing `elm-compiler` install. Crucially, cabal does not allow two different incarnations (compiled from different code and/or against different installs of dependencies) of the same package with the same version number.
* At this point, I can only throw away the sandbox.

As I understand, the `cabal add-source` is exactly for dealing with such situations, because it will not *install* some package version into the sandbox.

The build/rebuild times that @evancz experienced, I can only explain to myself with that something changed in the source files in the dependency directories. Maybe they were touched by some `git` command in the meantime?

Having to throw away the sandbox whenever something has changed in one of the tool directories will of course lead to the same recompilations.

In summary, the approach without `cabal add-source` may work fine for people just wanting to *install* Elm from source, not touching the source in any way. But for developers, I doubt it will work. (But maybe I'm wrong. Evan must have some workflow on his own machine that works for him.)

jvoigtlaender commented:
Actually, my previous comment was maybe too pessimistic. Instead of having to go with `cabal add-source`, it should be possible to go with what [build.sh](https://github.com/jvoigtlaender/elm-platform/blob/patch-1/versions/0.15.1/build.sh) does, including my fixes from https://github.com/elm-lang/elm-platform/pull/65, but then when a source change in one of the project directories was made, always make sure that `cabal` is called with all projects at once. Something like `cabal install ./elm-compiler ./elm-package ./elm-make ./elm-reactor ./elm-repl`. That will only do new compilation for stuff that has changed, but if something *has* changed in the `elm-compiler` subdirectory, it will make sure that the other stuff that depends on `elm-compiler` (which is all of them) will be relinked as well, thus avoiding that the sandbox gets into a broken stage. (Actually, the packages that produce only executables could be left out from that command, as long as one knows that one doesn't need those executables to be rebuilt right now.)

(All this here concerns "developer mode" only. For doing-an-install-just-once the linked script along with my linked fixes should just work.)

rehno-lindeque commented:
> Something like cabal install `./elm-compiler ./elm-package ./elm-make ./elm-reactor ./elm-repl`.

@jvoigtlaender unless I'm missing something this is the suggestion Apanatshka made [above](https://github.com/elm-lang/elm-platform/pull/55#issuecomment-130517862). (It would be very easy to amend this PR to do that.)

evancz commented:
I have a workflow that works well with `--sandbox`. I never have to throw things away. My builds are as small as possible. I'll share that in the makefile I create.

evancz commented on installers/BuildFromSource.hs:
```diff
@@ -127,22 +127,32 @@ makeRepos artifactDirectory repos =
+    cabal ([ "install", "-j", "--bindir=../bin", "--ghc-options=\"-XFlexibleContexts\"" ] ++ filter (/= "elm-reactor") (map fst repos))
+
+    -- elm-reactor needs to be installed last because of a post-build dependency on elm-make
+    cabal [ "install", "-j", "--bindir=../", "elm-reactor" ]
```
How come this uses `--bindir=../` and the other one uses `--bindir=../bin`?

rehno-lindeque replied:
Oh sorry, that must be a mistake, it should definitely be `../bin`

evancz commented:
Alright, I have convinced myself that this is probably the best way to go right now. I don't think I'll develop this way, but if it works better for linux people who just want to use Elm, I am down.

Once the `--bindir` and FlexibleInstances questions are sorted, I'll merge this in and let people know.

evancz commented on installers/BuildFromSource.hs:
```diff
@@ -127,22 +127,32 @@ makeRepos artifactDirectory repos =
+    -- install all of the packages together in order to resolve transitive dependencies robustly
+    -- (install the dependencies a bit more quietly than the elm packages)
+    cabal ([ "install", "-j", "--only-dependencies", "--ghc-options=\"-w\"" ] ++ map fst repos)
+    cabal ([ "install", "-j", "--bindir=../bin", "--ghc-options=\"-XFlexibleContexts\"" ] ++ filter (/= "elm-reactor") (map fst repos))
```
Why do we need -XFlexibleContexts here? Is it to get things working with 7.10?

Apanatshka replied:
I suspect that's the reason. I've needed `-XFlexibleContexts` for a while now when building some part of Elm on my laptop (GHC 7.10, because Arch bleeding edge).

Apanatshka replied:
Oh, it's definitely GHC 7.10, it says so in the [commit where it's introduced](https://github.com/rehno-lindeque/elm-platform/commit/ef7511453a8e45c48b0c3b80b31e9cfde1790d1f)

rehno-lindeque replied:
Yes, tons of packages broke with 7.10 due to FlexibleContexts being required. (it could be fixed in source files intead, but... this was faster)

jvoigtlaender commented:
@rehno-lindeque, yes, what I suggested above was what @Apanatshka suggested.

rehno-lindeque commented:
Ok, I rebased this and fixed `bindir` for `elm-reactor`. The [improvement above](https://github.com/elm-lang/elm-platform/pull/55#issuecomment-130517862) would be very easy to apply if someone wants to take a shot at it (I mostly don't want to rebuild the sandbox from scratch again because that's the time-consuming part :))

evancz commented:
Cool, let's see what happens! Thanks for sticking with this. This is some of the hardest stuff because everything takes like an hour to even try out and it mostly breaks, so I really appreciate your work here!

rehno-lindeque commented:
I'm glad I could help!
